storey/sources.py

```diff
@@ -622,69 +620,94 @@ def __init__(
+                        self.context.logger.error(str(key_error))
+
         return await self._do_downstream(_termination_obj)
 
+    def _get_by_field_or_index(self, field, body: OrderedDict, field_type: str, raise_exception=False):
+        result = body[field]
+        if raise_exception:
+            self._is_nan_validator(result=result, body=body, field_type=field_type, field=field)
+        return result
 
-class CSVSource(_IterableSource, WithUUID):
-    """
-    Reads CSV files as input source for a flow.
-
-    :parameter paths: paths to CSV files
-    :parameter header: whether CSV files have a header or not. Defaults to False.
-    :parameter build_dict: whether to format each record produced from the input file as a dictionary (as opposed to a
-        list). Default to False.
-    :parameter key_field: the CSV field to be used as the key for events. May be an int (field index) or string (field
-        name) if with_header is True. Defaults to None (no key). Can be a list of keys
-    :parameter time_field: the CSV field to be parsed as the timestamp for events. May be an int (field index) or string
-        (field name) if with_header is True. Defaults to None (no timestamp field).
-    :parameter timestamp_format: timestamp format as defined in datetime.strptime(). Default to ISO-8601 as defined in
-        datetime.fromisoformat().
-    :parameter id_field: the CSV field to be used as the ID for events. May be an int (field index) or string (field
-        name) if with_header is True. Defaults to None (random ID will be generated per event).
-    :parameter type_inference: Whether to infer data types from the data (when True), or read all fields in as strings
-        (when False). Defaults to True.
-    :parameter parse_dates: list of columns (names or integers) that will be attempted to parse as date column
+    def _field_validator(self, df, key_field, id_field, path="file path was not provided."):
+        df = df.reset_index()
+        if key_field:
+            key_field = [key_field] if not isinstance(key_field, list) else key_field
+            missing_keys = list(set(key_field) - set(df.columns))
+            if missing_keys:
+                raise KeyError(f"KeyError occurred: keys {missing_keys} missing from df. Df path: {path}")
+        if id_field and id_field not in df.columns:
+            raise KeyError(f"KeyError occurred: id field '{id_field}' missing from df. Df path: {path}")
 
-    for additional params, see documentation of  :class:`~storey.flow.Flow`
+    class NoneKeyException(Exception):
+        pass
+
+    def _is_nan_validator(self, result, body, field_type, field):
+        if pandas.isna(result) or result is None:
+            raise self.NoneKeyException(f"For {body} value of {field_type} {field} is None")
+
+
+class CSVSource(DataframeSource):
+    """Reads Csv files as input source for a flow.
+
+    :parameter paths: paths to Parquet files
```

Comment:
Parquet? What happened to the original documentation? This whole docstring looks completely wrong.